### PR TITLE
Refactored REST API common code, NFS and iSCSI drivers.

### PR DESCRIPTION
**Pull request by deiter**
- Added pagination support.
- Added configuration parameters for REST API connect/read
  timeouts, connection retries and backoff factor.
- Fixed HA failover.
- Added retries on EBUSY errors.
- Fixed HTTP authentication.
- Disabled non-blocking mandatory locks.

pep8
```
pep8 run-test-pre: PYTHONHASHSEED='1540489231'
pep8 runtests: commands[0] | flake8 .
pep8 runtests: commands[1] | /opt/stack/cinder/tools/config/check_uptodate.sh
pep8 runtests: commands[2] | /opt/stack/cinder/tools/check_exec.py /opt/stack/cinder/cinder /opt/stack/cinder/doc/source/ /opt/stack/cinder/releasenotes/notes
___________________________________ summary ____________________________________
  pep8: commands succeeded
  congratulations :)
```

Tempest iSCSI
```
======
Totals
======
Ran: 273 tests in 4389.0000 sec.
 - Passed: 267
 - Skipped: 6
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 3492.6979 sec.
```

Tempest NFS
```
======
Totals
======
Ran: 273 tests in 7428.0000 sec.
 - Passed: 266
 - Skipped: 7
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 6033.1749 sec.
```

pylint
```
stack@openstack-master-ns5-iscsi:~/cinder$ pylint cinder/volume/drivers/nexenta/ns5
Using config file /opt/stack/cinder/.pylintrc

------------------------------------
Your code has been rated at 10.00/10
```

### CODE REVIEW

File: cinder/volume/drivers/nexenta/options.py [hunk 1]
```diff
@@ -63,31 +61,53 @@
...
+               help='HTTP(S) port to connect to NexentaStor management '
+                    'REST API server. If it is equal zero, 8443 for '
+                    'HTTPS and 8080 for HTTP is used'),
     cfg.StrOpt('nexenta_rest_protocol',
                default='auto',
                choices=['http', 'https', 'auto'],
-               help='Use http or https for REST connection (default auto)'),
+               help='Use http or https for NexentaStor management '
+                    'REST API connection (default auto)'),
+    cfg.FloatOpt('nexenta_rest_connect_timeout',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]: why float not int?

Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]:
```suggestion
    cfg.IntOpt('nexenta_rest_connect_timeout',
```

File: cinder/volume/drivers/nexenta/options.py [hunk 2]
```diff
@@ -63,31 +61,53 @@
...
+                 default=30,
+                 help='Specifies the time limit (in seconds), within '
+                      'which the connection to NexentaStor management '
+                      'REST API server must be established'),
+    cfg.FloatOpt('nexenta_rest_read_timeout',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]: same

File: cinder/volume/drivers/nexenta/options.py [hunk 3]
```diff
@@ -63,31 +61,53 @@
...
+                 default=300,
+                 help='Specifies the time limit (in seconds), '
+                      'within which NexentaStor management '
+                      'REST API server must send a response'),
+    cfg.FloatOpt('nexenta_rest_backoff_factor',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]: same

File: cinder/volume/drivers/nexenta/options.py [hunk 4]
```diff
@@ -122,22 +144,22 @@
                help='Group of hosts which are allowed to access volumes'),
     cfg.IntOpt('nexenta_iscsi_target_portal_port',
                default=3260,
-               help='Nexenta target portal port'),
+               help='NexentaStor target portal port'),
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]: this is also used in NEdge

Comment (Qeas) [inline cinder/volume/drivers/nexenta/options.py]:
```suggestion
               help='Nexenta appliance iSCSI target portal port'),
```

File: cinder/volume/drivers/nexenta/options.py (hunk 1 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/options.py]: According to the urllib3 doc: https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.timeout

total (integer, float, or None) 
connect (integer, float, or None)
read (integer, float, or None)

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 5]
```diff
@@ -14,330 +14,573 @@
...
 
+class NefException(VolumeDriverException):
+    def __init__(self, message=None, **kwargs):
+        defaults = {
+            'source': os.path.basename(__file__),
+            'name': 'NexentaCinderDriver',
+            'code': 'EBADMSG',
+            'message': 'UnknownError'
+        }
+        if message:
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: is this needed?

Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
```

File: cinder/volume/drivers/nexenta/options.py (hunk 4 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/options.py]: Ah ok, fixed!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 6]
```diff
@@ -14,330 +14,573 @@
...
+                raise NefException(content)
+            self.proxy.delay(self.stat[response.status_code])
+            LOG.debug('Retry %(text)s after delay',
+                      {'text': initial_text})
+            return self.request(self.method, self.path, **self.payload)
+        elif response.status_code == requests.codes.accepted:
+            path = self.getpath(content, 'monitor')
+            if not path:
+                message = (_('There is no monitor path '
+                             'is available for %(text)s')
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
                             'available for %(text)s')
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 7]
```diff
@@ -14,330 +14,573 @@
...
+                          'is no JSON data is available',
+                          {'text': text})
+                return response
+            LOG.debug('Append %(count)s data items to response',
+                      {'count': len(content['data'])})
+            self.data += content['data']
+            path = self.getpath(content, 'next')
+            if not path:
+                LOG.debug('Hook done on %(text)s: there '
+                          'is no next path is available',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
                          'is no next path available',
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 8]
```diff
@@ -14,330 +14,573 @@
...
+                message = (_('There is no monitor path '
+                             'is available for %(text)s')
+                           % {'text': text})
+                raise NefException({'code': 'ENOMSG', 'message': message})
+            self.proxy.delay(self.stat[response.status_code])
+            return self.request(method, path)
+        elif response.status_code == requests.codes.ok:
+            if not (isinstance(content, dict) and 'data' in content):
+                LOG.debug('Hook done on %(text)s: there '
+                          'is no JSON data is available',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
                          'is no JSON data available',
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 9]
```diff
@@ -14,330 +14,573 @@
...
+            if not path:
+                LOG.debug('Hook done on %(text)s: there '
+                          'is no next path is available',
+                          {'text': text})
+                return response
+            LOG.debug('Perform next session request %(method)s %(path)s',
+                      {'method': method, 'path': path})
+            return self.request(method, path)
+        LOG.debug('Hook done on %(text)s and '
+                  'return original response',
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
                  'returned original response',
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 10]
```diff
@@ -14,330 +14,573 @@
...
+        path = 'auth/login'
+        payload = {'username': self.proxy.username,
+                   'password': self.proxy.password}
+        data = json.dumps(payload)
+        kwargs = {'data': data}
+        self.proxy.delete_bearer()
+        response = self.request(method, path, **kwargs)
+        content = json.loads(response.content)
+        if not (isinstance(content, dict) and 'token' in content):
+            message = (_('There is no authentication token is available '
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
            message = (_('There is no authentication token available '
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py [hunk 11]
```diff
@@ -14,330 +14,573 @@
...
+    subj = 'service'
+    root = '/services'
+
+
+class NefNfs(NefCollections):
+    subj = 'NFS'
+    root = '/nas/nfs'
+
+
+class NefSmb(NefCollections):
```
Comment (Qeas) [inline cinder/volume/drivers/nexenta/ns5/jsonrpc.py]:
```suggestion
```

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 6 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 7 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 8 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 9 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 10 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/jsonrpc.py (hunk 11 above)
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/jsonrpc.py]: Fixed, thank you!

File: cinder/volume/drivers/nexenta/ns5/iscsi.py [hunk 12]
```diff
@@ -77,34 +80,30 @@ def __init__(self, *args, **kwargs):
...
-        self.nef_port = self.configuration.nexenta_rest_port
-        self.nef_user = self.configuration.nexenta_user
-        self.nef_password = self.configuration.nexenta_password
-        self.storage_pool = self.configuration.nexenta_volume
+        self.pool = self.configuration.nexenta_volume
         self.volume_group = self.configuration.nexenta_volume_group
         self.portal_port = self.configuration.nexenta_iscsi_target_portal_port
         self.portals = self.configuration.nexenta_iscsi_target_portals
+        self.dataset_sparsed = self.configuration.nexenta_sparse
+        self.dataset_deduplication = self.configuration.nexenta_dataset_dedup
```
[the remark this answers is not in the document]
Comment (deiter) [reply cinder/volume/drivers/nexenta/ns5/iscsi.py]: Keep deduce options for compatibility

**Pull request merged**